File: gosdk-develop/postcard/deserializer.go
package postcard

import (
	"fmt"
	"math"
	"math/big"
	"unicode/utf8"
)

// TypeResolver resolves a type_tag into the byte range of its value.
// Implemented by provider.IDLTypeResolver; injected per-deserializer so
// multiple clients can decode with their own loaded IDLs concurrently.
type TypeResolver interface {
	DecodeResource(typeTag uint64, bytes []byte) (valueBytes []byte, remaining []byte, err error)
	DecodeEvent(typeTag uint64, bytes []byte) (eventBytes []byte, remaining []byte, err error)
}

type Deserializer struct {
	buffer       []byte
	offset       int
	typeResolver TypeResolver
}

func NewDeserializer(data []byte) *Deserializer {
	return &Deserializer{buffer: data}
}

// SetTypeResolver sets the type_tag resolver used by api.ReadAnySerializeValueWithTypeTag
// and api.DeserializeEventEntry during this deserialization.
func (d *Deserializer) SetTypeResolver(r TypeResolver) { d.typeResolver = r }

// TypeResolver returns the resolver set via SetTypeResolver, or nil.
func (d *Deserializer) TypeResolver() TypeResolver { return d.typeResolver }

func (d *Deserializer) Remaining() int {
	return len(d.buffer) - d.offset
}

func (d *Deserializer) DeserializeStr() (string, error) {
	bytes, err := d.DeserializeBytes()
	if err != nil {
		return "", err
	}
	if !utf8.Valid(bytes) {
		return "", fmt.Errorf("invalid UTF-8 string")
	}
	return string(bytes), nil
}

func (d *Deserializer) DeserializeBytes() ([]byte, error) {
	length, err := d.DeserializeU32()
	if err != nil {
		return nil, err
	}
	if int(length) > d.Remaining() {
		return nil, fmt.Errorf("bytes length %d exceeds remaining buffer %d", length, d.Remaining())
	}
	return d.DeserializeFixedBytes(int(length))
}

func (d *Deserializer) DeserializeFixedBytes(length int) ([]byte, error) {
	return d.read(length)
}

func (d *Deserializer) DeserializeBool() (bool, error) {
	value, err := d.DeserializeU8()
	if err != nil {
		return false, err
	}
	if value != 0 && value != 1 {
		return false, fmt.Errorf("invalid postcard boolean")
	}
	return value == 1, nil
}

func (d *Deserializer) DeserializeU8() (uint8, error) {
	if d.offset >= len(d.buffer) {
		return 0, fmt.Errorf("reached end of postcard buffer")
	}
	b := d.buffer[d.offset]
	d.offset++
	return b, nil
}

func (d *Deserializer) DeserializeU16() (uint16, error) {
	value, err := d.deserializeVarUint64(math.MaxUint16, "u16")
	return uint16(value), err
}

func (d *Deserializer) DeserializeU32() (uint32, error) {
	value, err := d.deserializeVarUint64(math.MaxUint32, "u32")
	return uint32(value), err
}

func (d *Deserializer) DeserializeU64() (uint64, error) {
	return d.deserializeVarUint64(math.MaxUint64, "u64")
}

func (d *Deserializer) DeserializeU128() (*big.Int, error) {
	return d.deserializeVarUintBig(maxU128, "u128")
}

func (d *Deserializer) DeserializeI8() (int8, error) {
	value, err := d.DeserializeU8()
	return int8(value), err
}

func (d *Deserializer) DeserializeI16() (int16, error) {
	value, err := d.DeserializeU16()
	return int16(value), err
}

func (d *Deserializer) DeserializeI32() (int32, error) {
	value, err := d.DeserializeU32()
	return int32(value), err
}

func (d *Deserializer) DeserializeI64() (int64, error) {
	value, err := d.DeserializeU64()
	return int64(value), err
}

func (d *Deserializer) DeserializeEnumVariant() (uint32, error) {
	return d.DeserializeU32()
}

func (d *Deserializer) AssertEnd() error {
	if d.Remaining() != 0 {
		return fmt.Errorf("%d trailing bytes", d.Remaining())
	}
	return nil
}

func (d *Deserializer) deserializeVarUint64(max uint64, name string) (uint64, error) {
	var value uint64
	var shift uint
	for i := 0; i < 19; i++ {
		if d.offset >= len(d.buffer) {
			return 0, fmt.Errorf("reached end of postcard buffer")
		}
		b := d.buffer[d.offset]
		d.offset++
		value |= uint64(b&0x7f) << shift
		if (b & 0x80) == 0 {
			if value > max {
				return 0, fmt.Errorf("%s overflow", name)
			}
			return value, nil
		}
		shift += 7
	}
	return 0, fmt.Errorf("%s varint is too long", name)
}

func (d *Deserializer) deserializeVarUintBig(max *big.Int, name string) (*big.Int, error) {
	value := big.NewInt(0)
	part := new(big.Int)
	for i := 0; i < 19; i++ {
		if d.offset >= len(d.buffer) {
			return nil, fmt.Errorf("reached end of postcard buffer")
		}
		b := d.buffer[d.offset]
		d.offset++
		part.SetInt64(int64(b & 0x7f))
		part.Lsh(part, uint(i*7))
		value.Or(value, part)
		if (b & 0x80) == 0 {
			if value.Cmp(max) > 0 {
				return nil, fmt.Errorf("%s overflow", name)
			}
			return value, nil
		}
	}
	return nil, fmt.Errorf("%s varint is too long", name)
}

func (d *Deserializer) read(length int) ([]byte, error) {
	if length < 0 {
		return nil, fmt.Errorf("invalid read length")
	}
	if d.offset+length > len(d.buffer) {
		return nil, fmt.Errorf("reached end of postcard buffer")
	}
	result := make([]byte, length)
	copy(result, d.buffer[d.offset:d.offset+length])
	d.offset += length
	return result, nil
}

func (d *Deserializer) Peek(n int) ([]byte, error) {
	if d.offset+n > len(d.buffer) {
		return nil, fmt.Errorf("not enough bytes to peek")
	}
	return d.buffer[d.offset : d.offset+n], nil
}

func (d *Deserializer) Offset() int {
	return d.offset
}

func (d *Deserializer) Buffer() []byte {
	return d.buffer
}

func (d *Deserializer) Advance(n int) error {
	if n < 0 {
		return fmt.Errorf("Advance with negative offset %d", n)
	}
	if d.offset+n > len(d.buffer) {
		return fmt.Errorf("Advance(%d) would exceed buffer length %d (current offset %d)", n, len(d.buffer), d.offset)
	}
	d.offset += n
	return nil
}

File: gosdk-develop/python/milon_sdk/postcard/deserializer.py
"""Postcard 反序列化器（字节级复刻 Go 版 postcard.Deserializer）。

- 变长整数读取：LE 7-bit 组，与序列化严格互逆。
- u8 为单字节原样读取；u128 为最多 19 字节的大整数变长读取。
- TypeResolver 注入：反序列化带 type_tag 的资源/事件时，按 type_tag 动态
  分发到已加载 IDL 的 DecodeResource/DecodeEvent，返回消耗的字节区间。
"""
from __future__ import annotations

from typing import Callable, Optional, Protocol, TypeVar

U128_MAX = (1 << 128) - 1

T = TypeVar("T")


class TypeResolver(Protocol):
    """把 type_tag 解析为值字节区间（对应 Go postcard.TypeResolver）。"""

    def decode_resource(self, type_tag: int, data: bytes) -> tuple[bytes, bytes]:
        """返回 (值字节, 剩余字节)。"""
        ...

    def decode_event(self, type_tag: int, data: bytes) -> tuple[bytes, bytes]:
        """返回 (事件字节, 剩余字节)。"""
        ...


class Deserializer:
    __slots__ = ("_buf", "_offset", "type_resolver")

    def __init__(self, data: bytes | bytearray) -> None:
        self._buf = bytes(data)
        self._offset = 0
        self.type_resolver: Optional[TypeResolver] = None

    # ---------------------------------------------------------- 基础
    @property
    def buffer(self) -> bytes:
        return self._buf

    def offset(self) -> int:
        return self._offset

    def remaining(self) -> int:
        return len(self._buf) - self._offset

    def peek(self, n: int) -> bytes:
        if self._offset + n > len(self._buf):
            raise ValueError(f"not enough bytes to peek (need {n}, have {self.remaining()})")
        return self._buf[self._offset:self._offset + n]

    def advance(self, n: int) -> None:
        if n < 0:
            raise ValueError(f"Advance with negative offset {n}")
        if self._offset + n > len(self._buf):
            raise ValueError(
                f"Advance({n}) would exceed buffer length {len(self._buf)} "
                f"(current offset {self._offset})"
            )
        self._offset += n

    def read(self, length: int) -> bytes:
        if length < 0:
            raise ValueError("invalid read length")
        if self._offset + length > len(self._buf):
            raise ValueError("reached end of postcard buffer")
        result = self._buf[self._offset:self._offset + length]
        self._offset += length
        return result

    def assert_end(self) -> None:
        if self.remaining() != 0:
            raise ValueError(f"{self.remaining()} trailing bytes")

    # ---------------------------------------------------------- 变长整数
    def _deserialize_varuint64(self, max_value: int, name: str) -> int:
        value = 0
        shift = 0
        for _ in range(19):
            if self._offset >= len(self._buf):
                raise ValueError("reached end of postcard buffer")
            b = self._buf[self._offset]
            self._offset += 1
            value |= (b & 0x7F) << shift
            if (b & 0x80) == 0:
                if value > max_value:
                    raise ValueError(f"{name} overflow")
                return value
            shift += 7
        raise ValueError(f"{name} varint is too long")

    def _deserialize_varuint_big(self, max_value: int, name: str) -> int:
        value = 0
        for i in range(19):
            if self._offset >= len(self._buf):
                raise ValueError("reached end of postcard buffer")
            b = self._buf[self._offset]
            self._offset += 1
            value |= (b & 0x7F) << (i * 7)
            if (b & 0x80) == 0:
                if value > max_value:
                    raise ValueError(f"{name} overflow")
                return value
        raise ValueError(f"{name} varint is too long")

    # ---------------------------------------------------------- 具体类型
    def deserialize_u8(self) -> int:
        return self.read(1)[0]

    def deserialize_u16(self) -> int:
        return self._deserialize_varuint64(0xFFFF, "u16")

    def deserialize_u32(self) -> int:
        return self._deserialize_varuint64(0xFFFFFFFF, "u32")

    def deserialize_u64(self) -> int:
        return self._deserialize_varuint64(0xFFFFFFFFFFFFFFFF, "u64")

    def deserialize_u128(self) -> int:
        return self._deserialize_varuint_big(U128_MAX, "u128")

    def deserialize_i8(self) -> int:
        v = self.deserialize_u8()
        return v - 256 if v >= 128 else v

    def deserialize_i16(self) -> int:
        v = self._deserialize_varuint64(0xFFFF, "i16")  # Go 按 u16 读后强转 int16
        return v - 0x10000 if v >= 0x8000 else v

    def deserialize_i32(self) -> int:
        v = self._deserialize_varuint64(0xFFFFFFFF, "i32")
        return v - 0x100000000 if v >= 0x80000000 else v

    def deserialize_i64(self) -> int:
        v = self._deserialize_varuint64(0xFFFFFFFFFFFFFFFF, "i64")
        return v - 0x10000000000000000 if v >= 0x8000000000000000 else v

    def deserialize_bool(self) -> bool:
        value = self.deserialize_u8()
        if value not in (0, 1):
            raise ValueError("invalid postcard boolean")
        return value == 1

    def deserialize_fixed_bytes(self, length: int) -> bytes:
        return self.read(length)

    def deserialize_bytes(self) -> bytes:
        length = self.deserialize_u32()
        if length > self.remaining():
            raise ValueError(f"bytes length {length} exceeds remaining buffer {self.remaining()}")
        return self.deserialize_fixed_bytes(length)

    def deserialize_str(self) -> str:
        raw = self.deserialize_bytes()
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError as exc:  # Go 侧显式校验 UTF-8
            raise ValueError(f"invalid UTF-8 string: {exc}")

    def deserialize_enum_variant(self) -> int:
        return self.deserialize_u32()

    # ---------------------------------------------------------- 组合类型
    def deserialize_option(self, fn: Callable[["Deserializer"], T]) -> Optional[T]:
        has_value = self.deserialize_bool()
        if not has_value:
            return None
        return fn(self)

    def deserialize_seq(self, fn: Callable[["Deserializer"], T]) -> list[T]:
        length = self.deserialize_u32()
        if length > self.remaining():
            raise ValueError(f"seq length {length} exceeds remaining buffer {self.remaining()}")
        return [fn(self) for _ in range(length)]


def deserialize_postcard(
    data: bytes,
    fn: Callable[[Deserializer], T],
    allow_trailing: bool = False,
    resolver: Optional[TypeResolver] = None,
) -> T:
    """入口：反序列化并（默认）校验无尾随字节。resolver 注入 type_tag 解析器。"""
    d = Deserializer(data)
    d.type_resolver = resolver
    value = fn(d)
    if not allow_trailing:
        d.assert_end()
    return value
